File: tests/vistk/test/test_test.cxx
/*ckwg +5
 * Copyright 2012-2013 by Kitware, Inc. All Rights Reserved. Please refer to
 * KITWARE_LICENSE.TXT for licensing information, or contact General Counsel,
 * Kitware, Inc., 28 Corporate Drive, Clifton Park, NY 12065.
 */

#include <test_common.h>

#include <vistk/pipeline/utils.h>

#include <iostream>
#include <stdexcept>

#define TEST_ARGS ()

DECLARE_TEST(return_code);
DECLARE_TEST(error_string);
DECLARE_TEST(error_string_mid);
DECLARE_TEST(error_string_stdout);
DECLARE_TEST(error_string_second_line);
DECLARE_TEST(expected_exception);
DECLARE_TEST(unexpected_exception);
DECLARE_TEST(environment);

int
main(int argc, char* argv[])
{
  CHECK_ARGS(1);

  testname_t const testname = argv[1];

  DECLARE_TEST_MAP(tests);

  ADD_TEST(tests, return_code);
  ADD_TEST(tests, error_string);
  ADD_TEST(tests, error_string_mid);
  ADD_TEST(tests, error_string_stdout);
  ADD_TEST(tests, error_string_second_line);
  ADD_TEST(tests, expected_exception);
  ADD_TEST(tests, unexpected_exception);
  ADD_TEST(tests, environment);

  RUN_TEST(tests, testname);
}

IMPLEMENT_TEST(return_code)
{
  exit(EXIT_FAILURE);
}

IMPLEMENT_TEST(error_string)
{
  TEST_ERROR("an error");
}

IMPLEMENT_TEST(error_string_mid)
{
  std::cerr << "Test";
  TEST_ERROR("an error");
}

IMPLEMENT_TEST(error_string_stdout)
{
  std::cout << "Error: an error" << std::endl;
}

IMPLEMENT_TEST(error_string_second_line)
{
  std::cerr << "Not an error" << std::endl;
  TEST_ERROR("an error");
}

IMPLEMENT_TEST(expected_exception)
{
  EXPECT_EXCEPTION(std::logic_error,
                    throw std::logic_error("reason"),
                    "when throwing an exception");
}

IMPLEMENT_TEST(unexpected_exception)
{
  EXPECT_EXCEPTION(std::runtime_error,
                    throw std::logic_error("reason"),
                    "when throwing an unexpected exception");
}

IMPLEMENT_TEST(environment)
{
  vistk::envvar_name_t const envvar = "TEST_ENVVAR";

  vistk::envvar_value_t const envvalue = vistk::get_envvar(envvar);

  if (!envvalue)
  {
    TEST_ERROR("failed to get environment from CTest");
  }
  else
  {
    vistk::envvar_value_t const expected = vistk::envvar_value_t("test_value");

    if (*envvalue != *expected)
    {
      TEST_ERROR("Did not get expected value: "
                  "Expected: " << expected << " "
                  "Received: " << envvalue);
    }
  }
}

File: tests/bindings/python/test/test-test.py
#!@PYTHON_EXECUTABLE@
#ckwg +4
# Copyright 2012-2013 by Kitware, Inc. All Rights Reserved. Please refer to
# KITWARE_LICENSE.TXT for licensing information, or contact General Counsel,
# Kitware, Inc., 28 Corporate Drive, Clifton Park, NY 12065.


def test_return_code():
    import sys

    sys.exit(1)


def test_error_string():
    test_error('an error')


def test_error_string_mid():
    import sys

    sys.stderr.write('Test')
    test_error('an error')


def test_error_string_stdout():
    import sys

    sys.stdout.write('Error: an error\n')


def test_error_string_second_line():
    import sys

    sys.stderr.write('Not an error\n')
    test_error("an error")


def raise_exception():
    raise NotImplementedError


def test_expected_exception():
    expect_exception('when throwing an exception', NotImplementedError,
                     raise_exception)


def test_unexpected_exception():
    expect_exception('when throwing an unexpected exception', SyntaxError,
                     raise_exception)


if __name__ == '__main__':
    import os
    import sys

    if not len(sys.argv) == 4:
        test_error("Expected three arguments")
        sys.exit(1)

    testname = sys.argv[1]

    os.chdir(sys.argv[2])

    sys.path.append(sys.argv[3])

    tests = \
        { 'return_code': test_return_code
        , 'error_string': test_error_string
        , 'error_string_mid': test_error_string_mid
        , 'error_string_stdout': test_error_string_stdout
        , 'error_string_second_line': test_error_string_second_line
        , 'expected_exception': test_expected_exception
        , 'unexpected_exception': test_unexpected_exception
        }

    from vistk.test.test import *

    run_test(testname, tests)
